### DISCUSSION TITLE
Using SimpleBar with IDOM

### DISCUSSION THREAD

**Discussion by Archmonger** [idom-team/idom#536] [Question] [answered]
I am trying to import [simplebar](https://www.npmjs.com/package/simplebar-react), but it appears to prevent the component from registering properly.

```python
simplebar_react = idom.web.module_from_template(
    "react", "simplebar-react", resolve_exports=True
)
simplebar = idom.web.export(simplebar_react, "SimpleBar")
```

Any advice?

**Comment by rmorshea** [comment, marked as answer]
So imports of the form `import thing from "thing";` are using the default export of a module. This requires the, as yet unreleased, `exports_default` parameter for `module_from_template`.

**Reply by Archmonger** [reply]
Since this is still unreleased, is this syntax open for discussion?

I think this might be more cleanly & robustly handled within the `export` function.

For example

```python
bootstrap = idom.web.module_from_template("react", "react-bootstrap@5.1.3") 
ModalDialog = idom.web.export(bootstrap, "ModalDialog", from="ModalDialog")
# This is equivalent to...
# import ModalDialog from 'react-bootstrap/ModalDialog'
```

This would allow developers to not repeatedly re-run `module_from_template` for every single component part.

**Reply by rmorshea** [reply]
I like the idea of not needing to call it multiple times, though it'd need to be implemented slightly differently.

The problem is that the `from_template` part of that function is meant to imply that I'm actually constructing a Javascript module from a template and saving that file somewhere. So we need to know all the submodules of `react-bootstrap` you want at the time the template is constructed.

Also, knowing where to find the appropriate submodule isn't especially straightforward and depends on the particular CDN being used. I'm not actually sure where, for example, the `Button.js` file is on https://esm.sh (the default CDN at the moment) while on https://esm.run it can be found at the path `react-bootstrap@2.0.2/esm/Button.js`

**Reply by Archmonger** [reply]
How about pulling down everything (similar to NPM) and then just letting things be resolved within the import clause?
In fact... It might be easier to just have `idom.web.module_from_template` use NPM as a mandatory dependency.

**Reply by rmorshea** [reply]
IDOM actually [used to do this](https://github.com/idom-team/idom/blob/0.15.0/idom/client/manage.py#L69-L114), but using NPM in this way adds a bunch of complexity:

- Users now have to worry about installing NPM
- You also require users to perform some sort of build step to turn code that's probably CJS into ESM modules
- Now there's a whole JS build system you have to maintain.
- What happens if third party Python packages call `module_from_template` with a conflicting version of a package?
- How do third party Python packages declare that they require a build step?

Fortunately or unfortunately, relying on CDNs is just way simpler.

**Reply by Archmonger** [reply]
Well then it seems like the most robust solution would be essentially mimicking NPM's behavior in `module_from_template`. Compiling everything into a single JS bundle (which is how I'm assuming NPM operates) makes a lot more sense then trying to poke around CDN endpoints to find a specific filename.

Bundling in Python would not only allow us to do it at run-time, but also would also play really well into creating predetermined Django static files.

Someone must've tried to do this in the past... I'll try to hunt for a python package that we might be able to use as inspiration.

**Reply by rmorshea** [reply]
So NPM is just a package manager. It doesn't create JS bundles. There are various bundlers like Rollup and Webpack that are beasts unto themselves that you have to use to do that (I use Rollup in the widget repo template).

Perhaps IDOM could provide some independent utilities for creating these bundles similar to the way [`jupyter-packaging`](https://github.com/jupyter/jupyter-packaging) does, but I'm hesitant to reintroduce NPM to IDOM like this again.

**Reply by Archmonger** [reply]
Similar idea to webpack but one bundle per node package, rather than a master bundle.

Thinking about it now, this might not even be needed. From IDOM's side of things, it would be just as effective to construct **all possible** JS modules from a react template within `module_from_template`, then create a dictionary mapping of these.

Would allow the `export` syntax to operate as I mentioned above.

Let me know your thoughts on both of these options.

**Reply by rmorshea** [reply]
That's almost identical to what IDOM used to do. This idea of turning each dependency into a bundle is something that [Snowpack](https://www.snowpack.dev/) does.

We could make this sort of thing a separate tool, but I don't think I'm going to build it into the core of IDOM.
